lib/cloud/provisioning/awsactions/create_oidc_idp.go

```diff
@@ -0,0 +1,96 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package awsactions
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/gravitational/trace"
+
+	awslib "github.com/gravitational/teleport/lib/cloud/aws"
+	"github.com/gravitational/teleport/lib/cloud/provisioning"
+	"github.com/gravitational/teleport/lib/integrations/awsoidc/tags"
+)
+
+// OpenIDConnectProviderCreator can create an OpenID Connect Identity Provider
+// (OIDC IdP) in AWS IAM.
+type OpenIDConnectProviderCreator interface {
+	// CreateOpenIDConnectProvider creates an AWS IAM OIDC IdP.
+	CreateOpenIDConnectProvider(ctx context.Context, params *iam.CreateOpenIDConnectProviderInput, optFns ...func(*iam.Options)) (*iam.CreateOpenIDConnectProviderOutput, error)
+}
+
+// CreateOIDCProvider wraps a [OpenIDConnectPRoviderCreator] in a
+// [provisioning.Action] that creates an OIDC IdP in AWS IAM when invoked.
+func CreateOIDCProvider(
+	clt OpenIDConnectProviderCreator,
+	thumbprints []string,
+	issuerURL string,
+	clientIDs []string,
+	tags tags.AWSTags,
+) (*provisioning.Action, error) {
+	details, err := formatDetails(identityProviderDetails{
+		IssuerURL:   issuerURL,
+		ClientIDs:   clientIDs,
+		Thumbprints: thumbprints,
+		Tags:        tags,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	config := provisioning.ActionConfig{
+		Name:    "CreateOpenIDConnectIdentityProvider",
+		Summary: "Create an OpenID Connect identity provider in AWS IAM for your Teleport cluster",
+		Details: details,
+		RunnerFn: func(ctx context.Context) error {
+			slog.InfoContext(ctx, "Creating OpenID Connect identity provider",
+				"issuer_url", issuerURL,
+			)
+			_, err = clt.CreateOpenIDConnectProvider(ctx, &iam.CreateOpenIDConnectProviderInput{
+				ThumbprintList: thumbprints,
+				Url:            &issuerURL,
+				ClientIDList:   clientIDs,
+				Tags:           tags.ToIAMTags(),
+			})
+			if err != nil {
+				awsErr := awslib.ConvertIAMv2Error(err)
```

Comment:
>  It is unfortunate that we still need to convert the error here so we knew this client must come from the SDK.

Yeah, I couldn't think of a way to avoid peaking into the implementation unfortunately

> Any benefits in doing the interfaces per function though

I like how it makes it easy to compose the client interfaces with just the methods they actually use.